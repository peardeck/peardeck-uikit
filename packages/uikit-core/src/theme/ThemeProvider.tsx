import React, { ReactNode } from "react";
import { ThemeProvider as StyledThemeProvider } from "styled-components";
import { ThemedComponentProps, theme as defaultTheme } from "./theme";

export interface ThemeProviderProps extends ThemedComponentProps {
  children: ReactNode;
}

// TODO: Add global styles/reset.

export const ThemeProvider = ({
  theme = defaultTheme,
  children,
}: ThemeProviderProps): JSX.Element => (
  <StyledThemeProvider theme={theme}>
    <>{children}</>
  </StyledThemeProvider>
);
