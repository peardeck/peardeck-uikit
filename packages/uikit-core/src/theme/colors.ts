/**
 * Some standard color scales that can be used for general-purpose
 * styling or as a basis for theme colors.
 * These can be references like: `<Thing color="blue.5" />`
 */
export default {
  black: "#000",
  white: "#fff",
  gray: [
    "#eaeaea",
    "#cfcfcf",
    "#b5b5b5",
    "#9c9c9c",
    "#838383",
    "#6b6b6b",
    "#555555",
    "#3f3f3f",
    "#2a2a2a",
  ],
  blue: [
    "#deeef6",
    "#c3e0ef",
    "#a9d2e8",
    "#8ec4e1",
    "#65aed7",
    "#3087B7",
    "#29739c",
    "#225f81",
    "#1b4b66",
    "#102e3e",
  ],
  green: [
    "#f7ffda",
    "#eaff96",
    "#dcff52",
    "#ceff0e",
    "#a0c900",
    "#6A8500",
    "#4f6300",
    "#344100",
    "#263000",
    "#191f00",
  ],
  yellow: [
    "#ffff9c",
    "#f9f3c1",
    "#f6eda1",
    "#f5e98f",
    "#f1e163",
    "#eedb44",
    "#ebd525",
    "#dac414",
    "#bba811",
    "#9c8c0e",
  ],
  orange: [
    "#ffe1aa",
    "#ffd588",
    "#ffca66",
    "#ffbe44",
    "#ffb222",
    "#FFA600",
    "#dd9000",
    "#bb7a00",
    "#996400",
    "#774e00",
  ],
  red: [
    "#f4d1d1",
    "#edb6b6",
    "#e69b9b",
    "#DD7676",
    "#d96464",
    "#D24949",
    "#c83131",
    "#ac2b2b",
    "#912424",
    "#761d1d",
  ],
};
