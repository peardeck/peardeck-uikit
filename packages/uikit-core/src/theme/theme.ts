import colorNames from "./colors";

/**
 * https://styled-system.com/theme-specification
 */

const colors = {
  ...colorNames,
  background: colorNames.white,
  text: colorNames.black,
  primary: "#3087b7",
  secondary: "#698600",
  accent: "#D24949",
  pear: "#c9de55",
  muted: colorNames.gray[3],
  border: colorNames.gray[5],
};

export const theme = {
  mediaQueries: {},
  breakpoints: [48, 52, 64].map((value): string => `${value}rem`),
  space: [0, 4, 8, 16, 32, 64, 128, 256, 512],
  colors,
  fonts: {
    // headers:
    body: ["ProximaNovaSoft-Regular", "sans-serif"],
    monospace: [
      "Consolas",
      '"Liberation Mono"',
      "Menlo",
      "Courier",
      "monospace",
    ],
  },
  fontSizes: [12, 14, 16, 20, 24, 32, 48, 64, 72],
  fontWeights: {
    light: 300,
    normal: 400,
    medium: 500,
    semibold: 600,
    bold: 700,
    heavy: 800,
  },
  lineHeights: {
    condensedUltra: 1,
    condensed: 1.25,
    default: 1.5,

    body: 1.7,
    title: 1.25,
  },
  letterSpacings: [],
  sizes: [],
  borders: [0, `1px solid ${colors.border}`],
  borderStyles: {},
  borderWidths: [],
  radii: [0, 3, 6, 12, 24, 48, "100%"],
  shadows: {
    [-2]: "rgba(0,0,0,0.12) 0px 2px 4px inset",
    [-1]: "rgba(0,0,0,0.12) 0px 1px 2px inset",
    0: "none",
    1: "0 1px 4px 0 rgba(0,0,0,0.27)",
    2: "0 2px 8px 0 rgba(0,0,0,0.27)",
    3: "0 4px 16px 0 rgba(0,0,0,0.27)",
    4: "0 8px 32px 0 rgba(0,0,0,0.27)",

    button: "0 1px 4px 0 rgba(0,0,0,0.27)",
    buttonHover: "0 2px 8px 0 rgba(0,0,0,0.27)",
    buttonActive: "rgba(0,0,0,0.12) 0px 0px 1px inset",

    // button: "0 2px 8px 0 rgba(0,0,0,0.27)",
    // buttonHover: "0 2px 8px 0 rgba(0,0,0,0.27)",
    // buttonActive: "",
    // buttonFocus: "",

    // 1: "0 1px 2px 0 rgba(0,0,0,0.4)",
    // 2: "0 2px 4px 0 rgba(0,0,0,0.4)",
    // 3: "0 4px 8px 0 rgba(0,0,0,0.4)",
    // 4: "0 8px 16px 0 rgba(0,0,0,0.4)",
    // formControl: "rgba(27, 31, 35, 0.075) 0px 1px 2px inset",
    // formControlFocus: "rgba(3, 102, 214, 0.3) 0px 0px 0px 0.2em",
    // glow: ``,
  },
  zIndices: [undefined, 0, 1, 2, 10, 100, 200, 10000],

  buttonColors: {
    default: {
      color: "#000000",
      backgroundColor: "#ffffff",
      hoverBackgroundColor: "#f2f2f2",
      borderColor: "#cccccc",
    },
    primary: {
      color: "#ffffff",
      backgroundColor: "#3087b7",
      hoverBackgroundColor: "#2b78a3",
      borderColor: "#ffffff",
    },
    secondary: {
      color: "#ffffff",
      backgroundColor: "#698600",
      hoverBackgroundColor: "#556d00",
      borderColor: "#ffffff",
    },
    accent: {
      color: "#ffffff",
      backgroundColor: "#d24949",
      hoverBackgroundColor: "#ba2e2e",
      borderColor: "#ffffff",
    },
  },

  colorStyles: {
    success: {
      color: "#8FA340",
    },
    info: {
      color: "#444546",
    },
    danger: {
      color: "#D24949",
    },
    warning: {
      color: "#F5E98F",
    },
  },

  textStyles: {
    caps: {
      textTransform: "uppercase",
    },
    smallCaps: {
      textTransform: "uppercase",
      letterSpacing: "0.1em", // TODO: Replace with `letterSpacings` value
    },
  },
};

export type Theme = typeof theme;

export interface ThemedComponentProps {
  theme?: Theme;
}
