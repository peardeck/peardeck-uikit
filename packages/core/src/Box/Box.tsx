import styled from "styled-components";
import {
  color,
  ColorProps,
  layout,
  LayoutProps,
  space,
  SpaceProps,
} from "styled-system";
import theme, { ThemedComponentProps } from "../Theme/theme";

type BoxProps = ColorProps & LayoutProps & SpaceProps & ThemedComponentProps;

export const Box: React.FunctionComponent<BoxProps> = styled.div`
  ${color}
  ${layout}
  ${space}
`;

Box.defaultProps = { theme };
