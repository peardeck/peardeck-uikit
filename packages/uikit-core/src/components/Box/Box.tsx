import styled from "styled-components";
import {
  border,
  BorderProps,
  boxShadow,
  BoxShadowProps,
  color,
  ColorProps,
  flexbox,
  FlexboxProps,
  layout,
  LayoutProps,
  position,
  PositionProps,
  space,
  SpaceProps,
  ZIndexProps,
  zIndex,
} from "styled-system";
import { theme, ThemedComponentProps } from "../../theme/theme";

export type BoxProps = ThemedComponentProps &
  BorderProps &
  BoxShadowProps &
  ColorProps &
  LayoutProps &
  SpaceProps &
  FlexboxProps &
  PositionProps &
  ZIndexProps;

export const Box = styled.div<BoxProps>`
  ${border}
  ${boxShadow}
  ${color}
  ${layout}
  ${space}
  ${flexbox}
  ${position}
  ${zIndex}
`;

Box.defaultProps = { theme };

Box.displayName = "Box";
