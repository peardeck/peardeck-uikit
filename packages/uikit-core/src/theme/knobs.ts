import { number, select } from "@storybook/addon-knobs";
import { theme } from "./theme";

/**
 * This contains various predefined knobs and helper methods for creating
 * knobs.
 *
 * Notes:
 *
 * * The `createOptions...` methods are used to generated options from existing
 *   scales or arrays more easily, while additionally adding a "not specified"
 *   option to select, which effect means "the default value when the prop
 *   is not given"
 *
 * * The `select` knob does not display options with an underlying value of
 *   `undefined` - this is why the default and "not specified" values are
 *   actually set to `null`. Then each knob uses the `nullToUndefined`
 *   helper to convert it correctly to undefined.
 */

export const GROUP_ID_TYPOGRAPHY = "Typography";
export const GROUP_ID_COLOR = "Color";
export const GROUP_ID_SPACE = "Space";
export const GROUP_ID_DISPLAY = "Display";
export const GROUP_ID_BORDER = "Border";
export const GROUP_ID_LAYOUT = "Layout";

/* *************************************
 * Helper methods
 ************************************ */

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const nullToUndefined = (result: any) => (result === null ? undefined : result);

/**
 * Creates options for a `select` knob that also includes a "not specified"
 * value, with the option labels matching the array _values_.
 * @param values An array of values
 */
export const createOptionsFromArray = (values: (string | number)[]) =>
  values.reduce((options, value) => ({ ...options, [value]: value }), {
    "not specified": null,
  });

/**
 * Creates options for a `select` knob that also includes a "not specified"
 * value, with the option labels matching the array _indexes_.
 * @param values An array of values
 */
export const createOptionsFromScale = (values: (string | number)[]) =>
  values.reduce((options, _, index) => ({ ...options, [index]: index }), {
    "not specified": null,
  });

/**
 * Creates options for a `select` knob that also includes a "not specified"
 * value, with the option labels generated using the keys of the map
 * @param map A map of values
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const createOptionsFromMap = (map: { [key: string]: any }) =>
  Object.keys(map).reduce((options, key) => ({ ...options, [key]: key }), {
    "not specified": null,
  });

/* *************************************
 * Predefined knobs
 ************************************ */

/**
 * A knob for selecting a font size value from the theme-defined scale of font sizes scale
 */
export const themeFontSizeKnob = (label = "Font size", initial = null) =>
  nullToUndefined(
    select(
      label,
      createOptionsFromScale(theme.fontSizes),
      initial,
      GROUP_ID_TYPOGRAPHY
    )
  );

/**
 * A knob for selecting a font family value from the theme-defined font families
 */
export const themeFontFamilyKnob = (label = "Font Family", initial = null) =>
  nullToUndefined(
    select(
      label,
      createOptionsFromMap(theme.fonts),
      initial,
      GROUP_ID_TYPOGRAPHY
    )
  );

/**
 * A knob for selecting a font weight value from the theme-defined font weights
 */
export const themeFontWeightKnob = (label = "Font Weight", initial = null) =>
  nullToUndefined(
    select(
      label,
      createOptionsFromMap(theme.fontWeights),
      initial,
      GROUP_ID_TYPOGRAPHY
    )
  );

/**
 * A knob for selecting a line height value from the theme-defined line heights
 */
export const themeLineHeightKnob = (label = "Line Height", initial = null) =>
  nullToUndefined(
    select(
      label,
      createOptionsFromMap(theme.lineHeights),
      initial,
      GROUP_ID_TYPOGRAPHY
    )
  );

/**
 * A knob for selecting a color value from the theme-defined (named) colors
 * (For _any_ color, use the `color` knob provided by `@storybook/addon-knobs`)
 */
export const themeColorKnob = (
  label = "Color",
  initial: null | string = null
) =>
  nullToUndefined(
    select(label, createOptionsFromMap(theme.colors), initial, GROUP_ID_COLOR)
  );

/**
 * A knob for selecting an opacity value from 0 to 1.
 */
export const opacityKnob = (label = "Opacity", initial = 1) =>
  nullToUndefined(
    number(
      label,
      initial,
      { range: true, min: 0, max: 1, step: 0.01 },
      GROUP_ID_COLOR
    )
  );

/**
 * A knob for selecting a letter spacing value from the theme-defined letter spacings scale
 */
export const themeLetterSpacingKnob = (
  label = "Letter Spacing",
  initial = null
) =>
  nullToUndefined(
    select(
      label,
      createOptionsFromScale(theme.letterSpacings),
      initial,
      GROUP_ID_SPACE
    )
  );

/**
 * A knob for selecting a border value from the theme-defined borders
 */
export const themeBordersKnob = (label = "Border", initial = null) =>
  nullToUndefined(
    select(
      label,
      createOptionsFromScale(theme.borders),
      initial,
      GROUP_ID_BORDER
    )
  );

/**
 * A knob for selecting a radius value from the theme-defined radii scale
 */
export const themeRadiiKnob = (label = "Border Radius", initial = null) =>
  nullToUndefined(
    select(label, createOptionsFromScale(theme.radii), initial, GROUP_ID_BORDER)
  );

/**
 * A knob for selecting a shadow value from the theme-defined shadows scale
 */
export const themeShadowsKnob = (
  label = "Shadow",
  initial: number | null = null
) =>
  nullToUndefined(
    select(
      label,
      createOptionsFromScale(theme.shadows),
      initial,
      GROUP_ID_BORDER
    )
  );

/**
 * A knob for selecting a spacing value from the theme-defined space scale
 */
export const themeSpaceKnob = (label = "Spacing", initial = null) =>
  nullToUndefined(
    select(label, createOptionsFromScale(theme.space), initial, GROUP_ID_SPACE)
  );

/**
 * A knob for selecting a standard font style value
 */
export const fontStyleKnob = (label = "Font Style", initial = null) =>
  nullToUndefined(
    select(
      label,
      createOptionsFromArray(["normal", "italic", "oblique"]),
      initial,
      GROUP_ID_TYPOGRAPHY
    )
  );

/**
 * A knob for selecting a standard text alignment attribute value
 */
export const textAlignKnob = (label = "Text Align", initial = null) =>
  nullToUndefined(
    select(
      label,
      createOptionsFromArray([
        "inherit",
        "initial",
        "revert",
        "unset",
        "center",
        "end",
        "justify",
        "left",
        "right",
        "start",
      ]),
      initial,
      GROUP_ID_TYPOGRAPHY
    )
  );

/**
 * A knob for selecting a standard display attribute value
 */
export const displayKnob = (label = "Display", initial = null) =>
  nullToUndefined(
    select(
      label,
      createOptionsFromArray([
        "block",
        "inline",
        "run-in",
        "inline-block",
        "flex",
        "table",
        "grid",
        "list-item",
        "none",
      ]),
      initial,
      GROUP_ID_DISPLAY
    )
  );

export const elementTypes = ["span", "div", "p"];

/**
 * A knob for selecting an element type (for an "as" prop) from a predefined
 * list of element types. (feel free to add additional types here if needed)
 */
export const elementTypeKnob = (label = "Element Type", initial = null) =>
  nullToUndefined(
    select(label, createOptionsFromArray(["span", "div", "p"]), initial)
  );

/* *************************************
 * Knob groups
 ************************************ */

/**
 * A collection of knobs for changing topography-related values.
 */
export const typographyKnobs = () => ({
  fontFamily: themeFontFamilyKnob(),
  fontSize: themeFontSizeKnob(),
  fontWeight: themeFontWeightKnob(),
  lineHeight: themeLineHeightKnob(),
  letterSpacing: themeLetterSpacingKnob(),
  fontStyle: fontStyleKnob(),
  textAlign: textAlignKnob(),
});

/**
 * Creates a collection of knobs for changing space-related values.
 */
export const spaceKnobs = () => ({
  m: themeSpaceKnob("Margin"),
  p: themeSpaceKnob("Padding"),
});

/**
 * Creates a collection of knobs for changing color-related values.
 */
export const colorKnobs = () => ({
  color: themeColorKnob("Foreground Color"),
  backgroundColor: themeColorKnob("Background Color"),
  opacity: opacityKnob(),
});

/**
 * Creates a collection of knobs for changing display-related values.
 */
export const displayKnobs = () => ({
  display: displayKnob(),
});

export const borderKnobs = () => ({
  border: themeBordersKnob(),
  boxShadow: themeShadowsKnob("Box Shadow"),
  borderRadius: themeRadiiKnob("Border Radius"),
});

export const layoutKnobs = () => ({
  // width,
  // height,
  // minWidth,
  // minHeight,
  // maxWidth,
  // maxHeight,
  // size,
  // overflow,
  // display,
  // verticalAlign,
});
