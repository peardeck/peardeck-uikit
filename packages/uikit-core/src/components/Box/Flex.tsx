import { HTMLAttributes } from "react";
import styled from "styled-components";
import { theme } from "../../theme/theme";
import { Box, BoxProps } from "./Box";

export type BaseFlexProps = BoxProps;

export const Flex = styled(Box)<BaseFlexProps>``;

export type FlexProps = BaseFlexProps & HTMLAttributes<HTMLElement>;

Flex.defaultProps = {
  display: "flex",
  theme,
};

Flex.displayName = "Flex";
