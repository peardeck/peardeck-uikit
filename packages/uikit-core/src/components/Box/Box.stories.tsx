import React from "react";
import { text } from "@storybook/addon-knobs";
import { Box } from "./Box";
import { theme } from "../../theme/theme";
import { colorKnobs, spaceKnobs } from "../../theme/knobs";

export default {
  component: Box,
  title: "core|Box",
};

const contentKnob = (initial: string) => text("Content", initial);

export const Default = () => (
  <Box {...colorKnobs()} {...spaceKnobs()}>
    {contentKnob(
      "Use the Box component to control width, margin, padding, and color."
    )}
  </Box>
);

export const FiftyPercentWidth = () => (
  <Box width={1 / 2} bg="yellow">
    {contentKnob("This is a box with 50% width.")}
  </Box>
);

FiftyPercentWidth.story = {
  name: "50% Width",
};

export const Inline = () => (
  <Box>
    Boxes are block-level by default, but can also be{" "}
    <Box display="inline" bg="#f0f">
      inline
    </Box>
  </Box>
);

export const Responsive = () => (
  <Box
    width={[1, 3 / 4, 1 / 2]}
    bg={["#ffff00", "#ff0000", "#00aa00"]}
    color="black"
    p={1}
  >
    This example shows how to use breakpoints. Based on the breakpoints
    specified in our theme, the background color and width of this box will
    change as the window size changes:
    <table
      style={{
        padding: "1em",
        border: "1px solid black",
        borderCollapse: "collapse",
        width: "100%",
      }}
    >
      <thead>
        <tr>
          <th>Size</th>
          <th>Example device</th>
          <th>Width</th>
          <th>Background</th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>&gt;= {theme.breakpoints[2]}px</td>
          <td>Desktop</td>
          <td>50%</td>
          <td>Green</td>
        </tr>
        <tr>
          <td>&gt;= {theme.breakpoints[1]}px</td>
          <td>iPad</td>
          <td>75%</td>
          <td>Red</td>
        </tr>
        <tr>
          <td>&gt;= {theme.breakpoints[0]}px</td>
          <td>iPhone</td>
          <td>100%</td>
          <td>Yellow</td>
        </tr>
      </tbody>
    </table>
  </Box>
);
