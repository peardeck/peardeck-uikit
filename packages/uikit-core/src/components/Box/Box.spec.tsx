import "jest-styled-components";
import { create } from "react-test-renderer";
import React from "react";
import { Box } from "./Box";
import { theme } from "../../theme/theme";

describe("Box", () => {
  it("matches the snapshot", () => {
    const json = create(<Box />).toJSON();
    expect(json).toMatchSnapshot();
  });

  it("`size` prop sets width and height", () => {
    const json = create(<Box size={40} />).toJSON();
    expect(json).toMatchSnapshot();
    expect(json).toHaveStyleRule("width", "40px");
    expect(json).toHaveStyleRule("height", "40px");
  });

  test("`width` prop sets width", () => {
    const json = create(<Box width={1 / 2} />).toJSON();
    expect(json).toMatchSnapshot();
    expect(json).toHaveStyleRule("width", "50%");
  });

  test("`m` prop sets margin", () => {
    const json = create(<Box m={2} />).toJSON();
    expect(json).toMatchSnapshot();
    expect(json).toHaveStyleRule("margin", `${theme.space[2]}px`);
  });

  test("`p` prop sets padding", () => {
    const json = create(<Box p={2} />).toJSON();
    expect(json).toMatchSnapshot();
    expect(json).toHaveStyleRule("padding", `${theme.space[2]}px`);
  });

  test("`color` prop sets color", () => {
    const json = create(<Box color="blue" />).toJSON();
    expect(json).toMatchSnapshot();
    expect(json).toHaveStyleRule("color", theme.colors.blue);
  });

  test("`bg` prop sets background color", () => {
    const json = create(<Box bg="green" />).toJSON();
    expect(json).toMatchSnapshot();
    expect(json).toHaveStyleRule("background-color", theme.colors.green);
  });
});
