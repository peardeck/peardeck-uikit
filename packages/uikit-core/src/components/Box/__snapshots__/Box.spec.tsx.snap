// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Box \`bg\` prop sets background color 1`] = `
.c0 {
  background-color: #191f00;
}

<div
  className="c0"
/>
`;

exports[`Box \`color\` prop sets color 1`] = `
.c0 {
  color: #3087B7;
}

<div
  className="c0"
  color="blue.5"
/>
`;

exports[`Box \`m\` prop sets margin 1`] = `
.c0 {
  margin: 8px;
}

<div
  className="c0"
/>
`;

exports[`Box \`p\` prop sets padding 1`] = `
.c0 {
  padding: 8px;
}

<div
  className="c0"
/>
`;

exports[`Box \`size\` prop sets width and height 1`] = `
.c0 {
  width: 40px;
  height: 40px;
}

<div
  className="c0"
  size={40}
/>
`;

exports[`Box \`width\` prop sets width 1`] = `
.c0 {
  width: 50%;
}

<div
  className="c0"
  width={0.5}
/>
`;

exports[`Box matches the snapshot 1`] = `
<div
  className=""
/>
`;
